### PR TITLE
[replay-ready-diagnostics] Patch 1: Add Failure_subkind taxonomy and classifier

### PR DESCRIPTION
**Pull request by subsetpark**
## Patch 1: Add Failure_subkind taxonomy and classifier

- Define type t = Ok | Auth_unavailable | Api_error of { status : int option } | Network_error | Timed_out | No_session_to_resume | Empty_response | Process_error | Other of string with [@@deriving show, eq, sexp_of, yojson].
- Define type init_info = { api_key_source : string option; model : string option; claude_code_version : string option } [@@deriving show, eq, yojson].
- Define classify ~classification ~init ~text_tail ~stderr_tail : t. Map Run_classification.{Process_error → Process_error; Timed_out → Timed_out; No_session_to_resume → No_session_to_resume; Success → Ok; Session_failed → sub-classify}.
- Session_failed sub-classification rules (first match wins): (1) text_tail contains 'Not logged in' OR 'Please run /login' → Auth_unavailable; (2) init.api_key_source = Some "none" AND no successful Final_result → Auth_unavailable; (3) stderr_tail matches /API Error: ([0-9]{3})/ → Api_error{status=Some N}; (4) stderr_tail contains getaddrinfo/EAI_AGAIN/ECONNREFUSED/TLS handshake → Network_error; (5) text_tail = "" AND stderr_tail = "" AND no Final_result → Empty_response; (6) otherwise → Other detail.
- Define to_string : t -> string for human-readable formatting (snake_case: 'ok', 'auth_unavailable', 'api_error_401', 'other').
- Add inline let%test blocks for each subkind classification rule plus a QCheck2 property: classify is total (never raises) over arbitrary inputs.

## Changes
- Define type t = Ok | Auth_unavailable | Api_error of { status : int option } | Network_error | Timed_out | No_session_to_resume | Empty_response | Process_error | Other of string with [@@deriving show, eq, sexp_of, yojson].
- Define type init_info = { api_key_source : string option; model : string option; claude_code_version : string option } [@@deriving show, eq, yojson].
- Define classify ~classification ~init ~text_tail ~stderr_tail : t. Map Run_classification.{Process_error → Process_error; Timed_out → Timed_out; No_session_to_resume → No_session_to_resume; Success → Ok; Session_failed → sub-classify}.
- Session_failed sub-classification rules (first match wins): (1) text_tail contains 'Not logged in' OR 'Please run /login' → Auth_unavailable; (2) init.api_key_source = Some "none" AND no successful Final_result → Auth_unavailable; (3) stderr_tail matches /API Error: ([0-9]{3})/ → Api_error{status=Some N}; (4) stderr_tail contains getaddrinfo/EAI_AGAIN/ECONNREFUSED/TLS handshake → Network_error; (5) text_tail = "" AND stderr_tail = "" AND no Final_result → Empty_response; (6) otherwise → Other detail.
- Define to_string : t -> string for human-readable formatting (snake_case: 'ok', 'auth_unavailable', 'api_error_401', 'other').
- Add inline let%test blocks for each subkind classification rule plus a QCheck2 property: classify is total (never raises) over arbitrary inputs.

## Files to Modify
- lib_core/failure_subkind.ml (create): Typed variant + classifier + yojson conversion + to_string. Pure module, no I/O.
- lib_core/failure_subkind.mli (create): Interface exposing the variant, classifier signature, yojson conversion, to_string.

## Gameplan Specification

```
module REPLAY_READY_DIAGNOSTICS.

> ══════════════════════════════════════════
> THE GUARANTEE
> ══════════════════════════════════════════
> Onton's observability is unified through a single typed
> Event variant routed via pluggable Sinks.  Pure decision
> (route, interested_in) is separated from effectful handling
> (consume, emit).  Every claude spawn produces a self-
> contained on-disk artifact directory at sessions/<uuid>/
> sufficient to diagnose the spawn's outcome offline.  Each
> spawn is classified by a typed Subkind enum.  The
> --upload-debug bundle includes a top-level manifest.json
> with a per-patch most-recent-subkind summary and no raw
> secrets.

Event.
Sink.
Spawn.
Meta.
Subkind.

ok => Subkind.
auth-unavailable => Subkind.
api-error => Subkind.
network-error => Subkind.
timed-out => Subkind.
no-session-to-resume => Subkind.
empty-response => Subkind.
process-error => Subkind.
other => Subkind.

> Telemetry layer.
sink-name s: Sink => String.
interested-in? s: Sink, e: Event => Bool.
consumed? s: Sink, e: Event => Bool.
route sinks: [Sink], e: Event => [Sink].

> Spawn / artifact layer.
spawn-completed? sp: Spawn => Bool.
spawn-uuid sp: Spawn => String.
spawn-meta sp: Spawn => Meta.
spawn-subkind sp: Spawn => Subkind.
artifact-dir-for sp: Spawn => String.
meta-path-for sp: Spawn => String.
schema-version m: Meta => Nat0.
subkind-of m: Meta => Subkind.
on-disk? p: String => Bool.
---

> Telemetry route is pure: a sink is in the output iff it was in the input AND interested.
all sinks: [Sink], e: Event, s: Sink, s in route sinks e | s in sinks.
all sinks: [Sink], e: Event, s: Sink, s in route sinks e | interested-in? s e.
all sinks: [Sink], e: Event, s: Sink, s in sinks, interested-in? s e | s in route sinks e.

> Emit dispatches an event to every interested sink (captured as consumed?).
all sinks: [Sink], e: Event, s: Sink, s in route sinks e | consumed? s e.

> Every completed spawn produces a finalized artifact dir on disk.
all sp: Spawn, spawn-completed? sp | on-disk? (artifact-dir-for sp).
all sp: Spawn, spawn-completed? sp | on-disk? (meta-path-for sp).

> meta.json carries schema_version = 1 and the classified subkind.
all sp: Spawn, spawn-completed? sp | schema-version (spawn-meta sp) = 1.
all sp: Spawn, spawn-completed? sp | subkind-of (spawn-meta sp) = spawn-subkind sp.

where

> ══════════════════════════════════════════
> Bundle manifest
> ══════════════════════════════════════════

Bundle.
Manifest.
File.

manifest-of b: Bundle => Manifest.
manifest-version m: Manifest => Nat0.
files-in b: Bundle => [File].
file-contents f: File => String.
contains-secret? s: String => Bool.
---

> Every --upload-debug bundle has a versioned manifest.
all b: Bundle | manifest-version (manifest-of b) = 1.

> Bundles never contain raw secrets.
all b: Bundle, f: File, f in files-in b | ~ contains-secret? (file-contents f).

```

## Patch Specification

```
module REPLAY_READY_DIAGNOSTICS_PATCH_1.

> Patch 1 introduces the typed failure-cause taxonomy.
> Pure module — no I/O, no state-machine change.

Classification.
Subkind.
InitInfo.
TextTail.
StderrTail.

process-error-class => Classification.
timed-out-class => Classification.
no-session-to-resume-class => Classification.
success-class => Classification.
session-failed-class => Classification.

ok => Subkind.
auth-unavailable => Subkind.
api-error => Subkind.
network-error => Subkind.
timed-out => Subkind.
no-session-to-resume => Subkind.
empty-response => Subkind.
process-error-sub => Subkind.
other => Subkind.

mentions-not-logged-in? t: TextTail => Bool.
mentions-api-error? e: StderrTail => Bool.
mentions-network-failure? e: StderrTail => Bool.
empty-text? t: TextTail => Bool.
empty-stderr? e: StderrTail => Bool.
api-key-source-none? i: InitInfo => Bool.

classify c: Classification, i: InitInfo, t: TextTail, e: StderrTail => Subkind.
---

> Direct mappings for non-Session_failed classifications.
all i: InitInfo, t: TextTail, e: StderrTail | classify success-class i t e = ok.
all i: InitInfo, t: TextTail, e: StderrTail | classify timed-out-class i t e = timed-out.
all i: InitInfo, t: TextTail, e: StderrTail | classify no-session-to-resume-class i t e = no-session-to-resume.
all i: InitInfo, t: TextTail, e: StderrTail | classify process-error-class i t e = process-error-sub.

> Session_failed Auth_unavailable rule (canonical first-match-wins).
all i: InitInfo, t: TextTail, e: StderrTail, mentions-not-logged-in? t | classify session-failed-class i t e = auth-unavailable.

```

## Implementation Notes

- The classifier is fully pure and total, but the `yojson` support is implemented manually rather than via `[@@deriving yojson]`. The requested variant shape `Api_error of { status : int option }` and the optional fields on `init_info` did not compile cleanly under the repo's current `ppx_yojson_conv` setup, so I exposed the same public conversion functions (`t_of_yojson` / `yojson_of_t`, `init_info_of_yojson` / `yojson_of_init_info`) with explicit total decoding instead.
- Manual decoders are intentionally forgiving: malformed `Failure_subkind` JSON decodes to `Other "invalid_failure_subkind"` and missing/non-string `init_info` fields decode to `None`. That keeps the new pure module aligned with the codebase rule that external-input decoders must not raise.
- The `api_key_source = Some "none"` rule is treated as a `Session_failed`-only heuristic. Patch 1 does not receive an explicit `saw_final_result` input, so the "no successful Final_result" part is satisfied structurally by only running the sub-classifier when `Run_classification` has already ruled the run out as `Success`.

---
## Summary by cubic
Add a pure `Failure_subkind` taxonomy and classifier to consistently label run outcomes, with JSON codecs and readable strings. Tightens `Session_failed` heuristics for auth and empty responses, and updates tests (including property-based) to ensure classification is total.

- **New Features**
  - New `Failure_subkind` module: enum (`Ok`, `Auth_unavailable`, `Api_error{status}`, `Network_error`, `Timed_out`, `No_session_to_resume`, `Empty_response`, `Process_error`, `Other`) and `init_info` (all fields optional). Pure.
  - JSON codecs are total: invalid subkind → `Other "invalid_failure_subkind"`; missing/non-string `init_info` fields → `None`.
  - `classify` maps `Run_classification` to subkinds with tightened `Session_failed` rules: auth hints or `api_key_source="none"` with empty tails → `Auth_unavailable`; “API Error: NNN” → `Api_error{status}`; network substrings → `Network_error`; both tails empty with empty or “(no error details)” → `Empty_response`; else `Other detail`.
  - `to_string` returns snake_case labels (e.g., `api_error_401`).
  - Tests: inline rule tests + QCheck property that `classify` never raises; generator uses valid `Success { stream_errors: string }` payloads.

<sup>Written for commit 5629a9a2613a53d0e663ece2b7dbef94ef06133a. Summary will update on new commits. <a href="https://cubic.dev/pr/flowglad/onton/pull/271?utm_source=github">Review in cubic</a></sup>

## Release Notes

* **Tests**
  * Added comprehensive property-based test coverage for failure classification scenarios, ensuring robust handling of various error conditions and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/flowglad/onton/pull/271?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: lib_core/failure_subkind.ml [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+let api_error_status stderr_tail =
+  match Re.exec_opt api_error_re stderr_tail with
+  | None -> None
+  | Some groups ->
+      Re.Group.get_opt groups 1 |> Option.bind ~f:Stdlib.int_of_string_opt
+
+let mentions_network_failure stderr_tail =
+  List.exists [ "getaddrinfo"; "EAI_AGAIN"; "ECONNREFUSED"; "TLS handshake" ]
+    ~f:(fun needle -> contains stderr_tail needle)
+
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/failure_subkind.ml:103]: Fixed. The `api_key_source = Some "none"` heuristic now only fires when both `text_tail` and `stderr_tail` are empty, so it no longer masks concrete downstream failures like `API Error: 401`. I also added a regression test covering the `api_key_source = none` + API-error case.

File: lib_core/failure_subkind.ml [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+let classify_session_failed ~init ~text_tail ~stderr_tail ~detail =
+  if mentions_not_logged_in text_tail then Auth_unavailable
+  else if api_key_source_none init then Auth_unavailable
+  else
+    match api_error_status stderr_tail with
+    | Some status -> Api_error { status = Some status }
+    | None when mentions_network_failure stderr_tail -> Network_error
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/failure_subkind.ml:113]: Fixed. The empty-tail branch now returns `Empty_response` only when the `Session_failed.detail` is also effectively empty (`"(no error details)"`), and otherwise falls through to `Other detail`. I added a regression test for the non-empty-detail + empty-tail case.

File: lib_core/failure_subkind.ml [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  let open Yojson.Safe.Util in
+  match json with
+  | `String "Ok" -> (Ok : t)
+  | `String "Auth_unavailable" -> Auth_unavailable
+  | `String "Network_error" -> Network_error
+  | `String "Timed_out" -> Timed_out
+  | `String "No_session_to_resume" -> No_session_to_resume
+  | `String "Empty_response" -> Empty_response
+  | `String "Process_error" -> Process_error
+  | `Assoc [ ("Api_error", payload) ] ->
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/failure_subkind.ml:31]: Fixed. `t_of_yojson` no longer uses `Yojson.Safe.Util.member` on an arbitrary payload value. The `Api_error` decoder now pattern-matches on `payload` being an assoc first and otherwise falls back to `status = None`, so malformed values like `null` or a string remain total and decode safely.

File: test/test_failure_subkind_properties.ml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+let gen_run_classification =
+  let open QCheck2.Gen in
+  oneof
+    [
+      map
+        (fun msg -> Run_classification.Process_error msg)
+        (string_small_of printable);
+      pure Run_classification.No_session_to_resume;
+      pure Run_classification.Timed_out;
+      map
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_failure_subkind_properties.ml:21]: Verified: `Run_classification.Success` is currently `Success of { stream_errors : string }` in `lib_core/run_classification.ml`, so the existing generator shape is correct. No code change was needed for this comment.

File: test/test_failure_subkind_properties.ml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        (fun stream_errors -> Run_classification.Success { stream_errors })
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_failure_subkind_properties.ml]: Verified against `lib_core/run_classification.mli`: `Run_classification.Success` is `Success of { stream_errors : string }`, so the generator payload shape was already correct. I still updated the generator to annotate `stream_errors` explicitly as `string` at the construction site so that the test makes the expected shape unambiguous in-code.

File: lib_core/failure_subkind.ml [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+  | Process_error -> `String "Process_error"
+  | Other detail -> `Assoc [ ("Other", `String detail) ]
+
+let init_info_of_yojson json =
+  let open Yojson.Safe.Util in
+  let read_opt_string field =
+    match member field json with
+    | `Null -> None
+    | value -> to_string_option value
+  in
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib_core/failure_subkind.ml:67]: Fixed. `init_info_of_yojson` now accepts only `\`String s` -> Some s`, `\`Null -> None`, and treats every other JSON shape as `None`. That removes the remaining `to_string_option` raise path and makes the decoder total on malformed non-string fields, matching the `lib_core` decoder rule.
